include/snap/plugin.h

```diff
@@ -23,43 +23,120 @@ limitations under the License.
 
 namespace Plugin {
 
+/**
+ * Type is the plugin type
+ */
 enum Type {
-  collector,
-  processor,
-  publisher
+  Collector,
+  Processor,
+  Publisher
 };
 
-class PluginInterface {
+/**
+ * RpcType is the rpc protocol the plugin should use.
+ * Only GRPC is supported for C++.
+ */
+enum RpcType {
+  GRPC = 2,
+};
 
+/**
+ * Strategy is the routing and caching strategy this plugin requires.
+ * A routing and caching Strategy is the method which snapd uses to cache 
+ * metrics, and select the correct running instance of a plugin.
+ */
+enum Strategy {
+  /**
+   * LRU: Least Recently Used.
+   * The default strategy.
+   */
+  LRU,
+
+  /**
+   * Sticky: Sticky Running plugins have a 1:1 relationship to Tasks.
+   */
+  Sticky,
+
+  /** ConfigBased: Bases routing decisions on the incoming config.
+   * The config based strategy hashes the incoming config data and
+   * uses it as a key to select a running plugin.  It can be useful for plugins
+   * which maintain a connection to a database, for instance.
+   */
+  ConfigBased
+};
+
+/**
+ * Meta is the metadata about the plugin.
+ */
+struct Meta final {
   public:
-    virtual Config::Policy getConfigPolicy() = 0;
+    Meta(Type type, std::string name, int version) : type(type),
+                                                     name(name),
+                                                     version(version) {};
+
+    Type type;
+    std::string name;
+    int version;
+
+    // These members all have defaults.
+    RpcType rpc_type = RpcType::GRPC;
```

Comment:
Looks like these local copies shadow the member variables? Are you actually setting them?